util/top_gen.py

```diff
@@ -4,11 +4,91 @@
 # SPDX-License-Identifier: Apache-2.0
 
 from mako.template import Template
+import yaml
 
-gpio_num = 5
-uart_num = 5
-i2c_num  = 2
-spi_num  = 7
+# First we parse the YAML input to get the parameters.
```

Comment:
We should decouple the pinmux generation from the top gen and make sure validation/parsing functionality is nicely separated.

As we've deprioritised top gen I'd focus entirely on a standalone pinmux tool that's written with reasonable modularity so at a later stage it could be used within topgen rather than run as a separate invocation.

Reply:
I've captured this here: https://github.com/lowRISC/sonata-system/issues/169